File: 02_Data_Structures/advanced_structures/fenwick_tree/implementations/fenwick_tree_basic.cpp
// /02_Data_Structures/advanced_structures/fenwick_tree/implementation/fenwick_tree_basic.cpp
// ------------------------------------------------------------------------------------------------
// Fenwick Tree (Binary Indexed Tree) - Basic Implementation in C++
//
// Problem Description:
//   Maintain an array A[1..n] supporting:
//     - Point updates: A[i] += delta
//     - Prefix sums:   sum(1..i)
//     - Range sums:    sum(l..r) = prefix(r) - prefix(l-1)
//
// Motivation:
//   Fenwick Tree gives O(log n) time for both updates and queries
//   while using only O(n) memory.
//
// Core Idea:
//   fenw[i] stores the sum of the range (i - LSB(i) + 1) ... i
//   where LSB(i) = i & -i.
//
// Algorithm Details:
//   update(i, delta):
//       while i <= n:
//           fenw[i] += delta
//           i += LSB(i)
//
//   prefix_sum(i):
//       while i > 0:
//           res += fenw[i]
//           i -= LSB(i)
//
// Complexity:
//   - update:      O(log n)
//   - prefix_sum:  O(log n)
//   - range_sum:   O(log n)
//   - space:       O(n)
//
// Variations included:
//   - 1-based index implementation
//   - Optional O(n) build constructor
//   - Optional find-by-prefix (order-statistics) support
//
// ------------------------------------------------------------------------------------------------

#include <bits/stdc++.h>
using namespace std;

class FenwickTree {
private:
    int n;
    vector<long long> fenw;

    // Least Significant Bit
    static int lsb(int x) {
        return x & -x;
    }

public:
    // Constructor: initializes tree with zeros
    FenwickTree(int n) : n(n), fenw(n + 1, 0) {}

    // Construct from vector (O(n) build)
    FenwickTree(const vector<long long>& arr) {
        n = arr.size();
        fenw.assign(n + 1, 0);

        // copy arr to fenw as 1-based
        for (int i = 1; i <= n; i++) fenw[i] = arr[i - 1];

        // build in O(n)
        for (int i = 1; i <= n; i++) {
            int j = i + lsb(i);
            if (j <= n)
                fenw[j] += fenw[i];
        }
    }

    // Point update: A[i] += delta
    void add(int i, long long delta) {
        while (i <= n) {
            fenw[i] += delta;
            i += lsb(i);
        }
    }

    // Prefix sum: sum A[1..i]
    long long prefix_sum(int i) const {
        long long res = 0;
        while (i > 0) {
            res += fenw[i];
            i -= lsb(i);
        }
        return res;
    }

    // Range sum: sum A[l..r]
    long long range_sum(int l, int r) const {
        if (l > r) return 0;
        return prefix_sum(r) - prefix_sum(l - 1);
    }

    // Order statistics:
    // Find smallest idx such that prefix_sum(idx) >= target
    // Requires all values non-negative.
    int find_by_prefix(long long target) const {
        if (target <= 0) return 1;

        long long total = prefix_sum(n);
        if (target > total) return -1; // not found

        int idx = 0;
        int bit = 1;

        // largest power of two <= n
        while (bit << 1 <= n) bit <<= 1;

        // binary lifting using Fenwick structure
        while (bit) {
            int next_idx = idx + bit;
            if (next_idx <= n && fenw[next_idx] < target) {
                target -= fenw[next_idx];
                idx = next_idx;
            }
            bit >>= 1;
        }
        return idx + 1; // convert to smallest valid index
    }
};

// ------------------------------------------------------------------------------------------------
// Optional: Basic demonstration and internal verification
// ------------------------------------------------------------------------------------------------
int main() {
    // Basic example
    vector<long long> arr = {3, 2, -1, 6, 5, 4, -3, 3, 7, 2, 3};

    FenwickTree ft(arr);  // O(n) build

    // Test prefix sums
    for (int i = 1; i <= (int)arr.size(); i++) {
        long long naive = accumulate(arr.begin(), arr.begin() + i, 0LL);
        assert(ft.prefix_sum(i) == naive);
    }

    // Range sum tests
    assert(ft.range_sum(2, 5) == accumulate(arr.begin() + 1, arr.begin() + 5, 0LL));
    assert(ft.range_sum(1, 1) == arr[0]);
    assert(ft.range_sum(4, 4) == arr[3]);

    // Point update tests
    ft.add(3, +10);  // A[3] += 10
    arr[2] += 10;
    long long p10 = accumulate(arr.begin(), arr.begin() + 10, 0LL);
    assert(ft.prefix_sum(10) == p10);

    // Order statistics on non-negative array
    vector<long long> pos = {1, 1, 2, 3, 5, 8};
    FenwickTree ft2(pos);

    assert(ft2.find_by_prefix(1) == 1);
    assert(ft2.find_by_prefix(3) == 3); // 1+1+2
    assert(ft2.find_by_prefix(4) == 4); // 1+1+2+3
    assert(ft2.find_by_prefix(20) == 6);
    assert(ft2.find_by_prefix(21) == -1);

    cout << "Fenwick Tree basic tests passed.\n";
    return 0;
}

File: 02_Data_Structures/advanced_structures/fenwick_tree/implementations/fenwick_tree_basic.py
# /02_Data_Structures/advanced_structures/fenwick_tree/implementation/fenwick_tree_basic.py
# -----------------------------------------------------------------------------
# Fenwick Tree (Binary Indexed Tree) - Basic Implementation
#
# Problem/Goal:
#   Maintain a dynamic array A[1..n] under:
#     1) Point updates: A[i] += delta
#     2) Prefix sums:   sum(1..i)
#     3) Optional range sums: sum(l..r) = sum(1..r) - sum(1..l-1)
#
# Why Fenwick Tree:
#   - O(log n) per update and query
#   - O(n) space
#   - Simpler than Segment Tree when only sums + point updates are needed
#
# Core Idea:
#   fenw[i] stores sum over (i - LSB(i) + 1) .. i, where LSB(i) = i & -i.
#   Query walks indices downward: i -= LSB(i).
#   Update walks indices upward:  i += LSB(i).
#
# Complexity:
#   - update/add:      O(log n)
#   - prefix_sum:      O(log n)
#   - range_sum:       O(log n)
#   - build (fast):    O(n)   (using internal parent linking)
#   - memory:          O(n)
#
# Variations Included:
#   - 1-based indexing (canonical)
#   - O(n) build from iterable
#   - range_sum(l, r)
#   - find_by_prefix(k): smallest idx such that prefix_sum(idx) >= k
#   - Thin 0-based index wrappers for convenience
#
# Interview Notes:
#   - Most implementations are 1-based; convert if input is 0-based.
#   - Mention LSB trick: i & -i
#   - Compare with Segment Tree (heavier but more flexible: min/max/range updates).
# -----------------------------------------------------------------------------


from typing import Iterable, List, Optional


class FenwickTree:
    """
    Fenwick Tree (Binary Indexed Tree) for prefix sums with point updates.

    Indexing:
        Internally 1-based. Public methods expect 1 <= i <= n.
        Convenience 0-based wrappers are provided (suffix `_0`).

    Attributes
    ----------
    n : int
        Size of the array.
    fenw : List[int/float]
        Internal tree array (length n+1), 1-based.

    Example
    -------
    >>> ft = FenwickTree(5)
    >>> ft.add(1, 5)     # A[1] += 5
    >>> ft.add(3, 2)     # A[3] += 2
    >>> ft.prefix_sum(3) # sum A[1..3] = 7
    7
    >>> ft.range_sum(2, 3)  # sum A[2..3] = 2
    2
    """

    def __init__(self, n: int):
        assert n >= 1, "FenwickTree size must be >= 1"
        self.n = n
        # fenw[0] is unused to keep arithmetic cleaner
        self.fenw: List[int] = [0] * (n + 1)

    # -----------------------
    # Helpers and core ops
    # -----------------------
    @staticmethod
    def _lsb(x: int) -> int:
        """Least Significant Bit (power of two that divides x)."""
        return x & -x

    def add(self, i: int, delta: int) -> None:
        """
        Point update: A[i] += delta  (1-based index)

        Time: O(log n)
        """
        while i <= self.n:
            self.fenw[i] += delta
            i += self._lsb(i)

    def prefix_sum(self, i: int) -> int:
        """
        Query prefix sum: sum A[1..i]  (1-based index)

        Time: O(log n)
        """
        res = 0
        while i > 0:
            res += self.fenw[i]
            i -= self._lsb(i)
        return res

    def range_sum(self, l: int, r: int) -> int:
        """
        Query range sum: sum A[l..r]  (1-based index, requires l <= r)

        Time: O(log n)
        """
        if l > r:
            return 0
        return self.prefix_sum(r) - self.prefix_sum(l - 1)

    # -----------------------
    # Construction
    # -----------------------
    @classmethod
    def from_iterable(cls, arr: Iterable[int]) -> "FenwickTree":
        """
        Build Fenwick Tree from 1D iterable in O(n).

        The trick:
            For i in [1..n], let j = i + LSB(i).
            Then fenw[j] += fenw[i], if j <= n.

        Time: O(n)
        """
        data = list(arr)
        n = len(data)
        assert n >= 1, "Input iterable must be non-empty"

        ft = cls(n)
        # copy as 1-based into fenw
        for i in range(1, n + 1):
            ft.fenw[i] = data[i - 1]
        # build in O(n)
        for i in range(1, n + 1):
            j = i + cls._lsb(i)
            if j <= n:
                ft.fenw[j] += ft.fenw[i]
        return ft

    # -----------------------
    # Order-statistics helper
    # -----------------------
    def find_by_prefix(self, target: int) -> Optional[int]:
        """
        Find smallest index idx in [1..n] such that prefix_sum(idx) >= target.
        Returns None if total sum < target.

        Assumptions:
            - All values are non-negative (monotone prefix sums).
        Time: O(log n)
        """
        if target <= 0:
            return 1
        # If total sum < target -> not found
        total = self.prefix_sum(self.n)
        if target > total:
            return None

        # Binary lifting over fenwick structure.
        idx = 0
        bit = 1
        # Largest power of two <= n
        while bit << 1 <= self.n:
            bit <<= 1

        # Descend using fenwick tree's implicit binary decomposition
        while bit:
            nxt = idx + bit
            if nxt <= self.n and self.fenw[nxt] < target:
                target -= self.fenw[nxt]
                idx = nxt
            bit >>= 1
        return idx + 1  # convert to 1-based smallest index

    # -----------------------
    # 0-based convenience API
    # -----------------------
    def add_0(self, i0: int, delta: int) -> None:
        """Point update for 0-based index i0."""
        self.add(i0 + 1, delta)

    def prefix_sum_0(self, i0: int) -> int:
        """Prefix sum over A[0..i0] for 0-based index i0."""
        return self.prefix_sum(i0 + 1)

    def range_sum_0(self, l0: int, r0: int) -> int:
        """Range sum over A[l0..r0] for 0-based indices."""
        return self.range_sum(l0 + 1, r0 + 1)


# -----------------------------------------------------------------------------
# Optional: simple self-check (can be removed in production)
# -----------------------------------------------------------------------------
if __name__ == "__main__":
    # Example usage and quick correctness check

    # Build from iterable (O(n))
    arr = [3, 2, -1, 6, 5, 4, -3, 3, 7, 2, 3]
    ft = FenwickTree.from_iterable(arr)

    # Check prefix sums vs naive
    for i in range(1, len(arr) + 1):
        naive = sum(arr[:i])
        assert ft.prefix_sum(i) == naive, f"prefix mismatch at {i}"

    # Range queries
    assert ft.range_sum(2, 5) == sum(arr[1:5])
    assert ft.range_sum(1, 1) == arr[0]
    assert ft.range_sum(4, 4) == arr[3]
    assert ft.range_sum(6, 3) == 0  # l > r

    # Point updates
    ft.add(3, +10)       # A[3] += 10
    arr[2] += 10
    assert ft.prefix_sum(10) == sum(arr[:10])

    # 0-based wrappers
    ft.add_0(0, -3)      # A[0] -= 3
    arr[0] -= 3
    assert ft.prefix_sum_0(4) == sum(arr[:5])
    assert ft.range_sum_0(2, 6) == sum(arr[2:7])

    # Order-statistics (find_by_prefix)
    # ensure non-negative for monotone prefix search
    arr_nonneg = [1, 1, 2, 3, 5, 8]
    ft2 = FenwickTree.from_iterable(arr_nonneg)
    # total = 20, find first idx with prefix >= k
    assert ft2.find_by_prefix(1) == 1
    assert ft2.find_by_prefix(3) == 3  # 1+1+2
    assert ft2.find_by_prefix(4) == 4  # 1+1+2+3
    assert ft2.find_by_prefix(20) == 6
    assert ft2.find_by_prefix(21) is None

    print("FenwickTree basic tests passed.")

